### PR TITLE
History indexing API

### PR DESCRIPTION
**Pull request by laerus**
History object supports indexing with some rules for extra functionality:
* index must be one of string, int, slice or tuple of length two
* an int returns the appropriate command in order
* a string returns the most recent command that contains the string
* a slice returns a list of commands
* if the index is a tuple:
    * the first item follows the previous
              three rules.
    * the second item is the slice of the arguments to be returned
 
see also bash history event and word designators
#1411 #1632

equivalents:

bash  | xonsh
-------- | --------
`!$`      | `__xonsh_history__[-1, -1]`
`!?cmd?:$` | `__xonsh_history__['cmd', -1]`
`!4:2-5` | `__xonsh_history__[4, 2:5]`

more functionality can be added to support more bash history idioms like
`^word^replace`.

### CODE REVIEW

**Comment by scopatz** [comment]
Looks like there is a flake8 error...

File: tests/test_history.py [hunk 1]
```diff
@@ -185,3 +185,19 @@ def test_parser_show(args, exp):
               'timestamp': False}
     ns = _hist_parse_args(shlex.split(args))
     assert ns.__dict__ == exp_ns
+
+
+# CMDS = ['ls', 'cat hello kitty', 'abc', 'def', 'touch me', 'grep from me']
```
Comment (scopatz) [inline tests/test_history.py]: Remove comment

File: xonsh/history.py [hunk 2]
```diff
@@ -605,16 +615,43 @@ def flush(self, at_exit=False):
         self.buffer.clear()
         return hf
 
-    def show(self, *args, **kwargs):
-        """Return shell history as a list
+    def _get(self):
```
Comment (scopatz) [inline xonsh/history.py]: This seems like an `__iter__` rather than a get.

**Comment by scopatz** [comment]
Thanks for putting this in @laerus!

**Comment by laerus** [comment]
still some minor typos on news, will fix that.

i was thinking about the `^replace^word` bashism and how i can add it in the indexing API,
would something like this `__xonsh_history__[-1, 'replace' : 'word']` be clear and intuitive?

**Comment by gforsyth** [comment]
> would something like this `__xonsh_history__[-1, 'replace' : 'word']` be clear and intuitive? 

how about `__xonsh_history__[-1].replace('old', 'new')` ?

**Comment by laerus** [comment]
@gforsyth yea it's an overkill mostly, just a little less verbose and it will also handle a list of commands. i'm fine with not implementing this, python is just an impulse to be creative.

**Comment by scopatz** [comment]
I agree with @gforsyth that replacement isn't exactly indexing and probably shouldn't be forced into that role.  The replacement bashism can still be implemented in a bashism xontrib.  In light of that, this looks good to go to me.  Any reason to delay merging this @laerus?

**Comment by laerus** [comment]
Maybe i should make `_cmd_filter` and `_args_filter` staticmethods? other than this no reason.

**Comment by scopatz** [comment]
Sure thing, if you want.  Just let me know when it is ready

**Comment by laerus** [comment]
this is ready!

**Comment by scopatz** [comment]
Looks good, thanks!

**Comment by laerus** [comment]
Thank you!

**Pull request merged**